dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/FilterParser.java

```diff
@@ -0,0 +1,273 @@
+import org.hisp.dhis.common.QueryFilter;
+import org.hisp.dhis.common.QueryOperator;
+import org.hisp.dhis.common.UID;
+import org.hisp.dhis.feedback.BadRequestException;
+
+/**
+ * Parser for the <code>filter</code> (<code>filterAttribute</code>) request parameter.
+ *
+ * <p><code>/changeLogs?filter</code> validation is implemented in {@link
+ * org.hisp.dhis.webapi.controller.tracker.RequestParamsValidator} and its parsing in its mapper.
+ * The requirements for these endpoints are slightly different. We should use this parser for
+ * changeLogs as soon as we need to support escaped values like dates which contain the segment
+ * separator ':'.
+ */
+public class FilterParser {
+  private FilterParser() {
+    throw new IllegalStateException("Utility class");
+  }
+
+  private static final char ESCAPE = '/';
+  private static final String ESCAPE_STRING = Character.toString(ESCAPE);
+  private static final String ESCAPED_ESCAPE = ESCAPE + ESCAPE_STRING;
+  private static final char COMMA = ',';
+  private static final String COMMA_STRING = Character.toString(COMMA);
+  private static final String ESCAPED_COMMA = ESCAPE + COMMA_STRING;
+  private static final char COLON = ':';
+  private static final String COLON_STRING = Character.toString(COLON);
+  private static final String ESCAPED_COLON = ESCAPE + COLON_STRING;
+
+  /**
+   * Parse given {@code input} string representing one or more filters for objects referenced by a
+   * UID like a tracked entity attribute or data element.
+   *
+   * <p>The expected input format is <br>
+   * <code>{uid}[:{operator}[:{value}]][,{uid}[:{operator}[:{value}]]]</code> <br>
+   * The value is mandatory and only allowed for binary operators. Multiple operator or
+   * operator:value pairs are allowed. A {@link QueryFilter} for each operator[:value] pair is added
+   * to the corresponding UID in the input order.
+   *
+   * <p>This method assumes that the input
+   *
+   * <ul>
+   *   <li>has been percent-decoded already
+   * </ul>
+   *
+   * @return accumulated {@link QueryFilter}s per UID
+   * @throws BadRequestException if the input is invalid
+   */
+  public static Map<UID, List<QueryFilter>> parseFilters(
+      @Nonnull String parameterName, @CheckForNull String input) throws BadRequestException {
+    try {
+      return parseFilters(input);
+    } catch (Exception e) {
+      throw new BadRequestException(parameterName + "=" + input + " is invalid. " + e.getMessage());
+    }
+  }
+
+  public static Map<UID, List<QueryFilter>> parseFilters(String input) throws BadRequestException {
+    Map<UID, List<QueryFilter>> result = new HashMap<>();
+    if (StringUtils.isBlank(input)) {
+      return result;
+    }
+
+    int start = 0;
+    int end = 0;
+    UID uid = null;
+    String operator = null;
+    String valueOrOperator = null;
+    int length = input.length();
+    while (end < length) {
+      int curChar = input.codePointAt(end);
+      int charCount = Character.charCount(curChar);
```

Comment:
I would add a comment explaining that `charCount` is needed to correctly deal with UNICODE chars.
Is it what is doing, right?